### PR TITLE
Trang - K33402

### PR DESCRIPTION
**Pull request by trangology**

### CODE REVIEW

File: src/main/scala/ru/ifmo/backend_2021/utils/SudokuUtils.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        val result = separator + stringRow(0) + stringRow(1) + stringRow(2)
+        result.replace("0", " ").stripTrailing() + "\n"
+    }
+
+    def isValidSudoku(rawSudoku: List[List[Int]]): Boolean = {
+        validateRows(rawSudoku) && validateColumns(rawSudoku) && validateBlocks(rawSudoku)
+    }
+
+    def renderSudoku(grid: List[List[Int]]): String = {
+        val ROW_INDICES = "  | 1 2 3 | 4 5 6 | 7 8 9 |"
```
Comment (IlyaHalsky) [inline src/main/scala/ru/ifmo/backend_2021/utils/SudokuUtils.scala:39]: Its better to keep consts outside of function, so they don't get evaluated every time

File: src/main/scala/ru/ifmo/backend_2021/game/Puzzle.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
+package ru.ifmo.backend_2021.game
+
+import ru.ifmo.backend_2021.utils._
+
+
+class Puzzle(val grid: List[List[Int]]) {
+
+    def this() = this(
```
Comment (IlyaHalsky) [inline src/main/scala/ru/ifmo/backend_2021/game/Puzzle.scala:8]: Could be replaced with default arguments to class argument.

File: src/main/scala/ru/ifmo/backend_2021/game/Game.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * @param puzzle the puzzle that contains a grid of 2-dimensional list of integers.
+ * @param state the current state of the game, may be `GET_STARTED`, `UNSOLVED`, `UNSOLVABLE`, or `SOLVED`.
+ */
+case class Game(player: Player, puzzle: Puzzle, state: State) {
+
+    def play(): Unit = {
+        this.state match {
+            case GET_STARTED => Game(this.player, puzzle, UNSOLVED).play()
+            case UNSOLVED =>
+                this.player match {
```
Comment (IlyaHalsky) [inline src/main/scala/ru/ifmo/backend_2021/game/Game.scala:17]: you don't have to write `this` everywhere

**Comment by IlyaHalsky** [comment]
35-7
Generally good code, suspiciously good. Take it as a compliment.

**Comment by trangology** [comment]
Thanks for the compliment - I think I deserve it. Also thanks for the review!
Just wondering why is there a minus while the formula for calculating the points is 27 + x?

**Comment by trangology** [comment]
And please check out the sheet. You put my points in the wrong row. My name is in row 29. Thank you!

**Comment by IlyaHalsky** [comment]
> Thanks for the compliment - I think I deserve it. Also thanks for the review!
> Just wondering why is there a minus while the formula for calculating the points is 27 + x?

I don't remember now, there was some logic.
Fixed points table.

**Comment by trangology** [comment]
That's nothing wrong with your calculation. I just don't feel like seeing the minus operator. 
You might not know how much effort I have put into this work. Anyway, thanks for the quick fix!

**Pull request closed without merging**